rad/monadic/diff-remote.rad

```diff
@@ -0,0 +1,226 @@
+                    (fn [t]
+                      (eq? (lookup :created-at t)
+                           (lookup :modified-at t))))]))
+
+(def validator/time-modified
+  (validator/keys
+   {:modified-at validator/timestamp}))
+
+(def validator/input
+  "Checks that the input as a valid new *chain input*, that is, it has an unused
+  nonce, is marked with this chain name, and is signed."
+  (validator/and
+   [(validator/keys input-keys)
+    validator/signed]))
+
+(def validator/diff-number
+  "A validator for numbers which refer to existing diffs."
+  (validator/pred "Existing diff number" (fn [n] (member? n (read-ref diffs)))))
+
+(def comment-keys
+  "A comment's body should be text that isn't too long."
+  {:body (validator/text 4000)})
+
+(def validator/full-comment
+  (validator/keys comment-keys))
+
+(def validator/diff-reference
+  "Refers to an existing issue."
+  (validator/optional-keys {:issue-number validator/diff-number}))
+
+(def validator/comment-input
+  "A comment input should refer to an existing issue."
+  (validator/and
+   [validator/input
+    validator/time-created
+    validator/full-comment
+    validator/diff-reference]))
+
+(def diff-keys
+  {:title       (validator/text 100)
+   :description (validator/text 4000)
+   :patch       (validator/type :string)
+   :state       (validator/member [:pending :retracted :accepted :rejected])
+   :comments    (validator/every validator/full-comment)})
+
+(def validator/full-diff (validator/keys diff-keys))
+
+(def validator/diff-input
+  "An diff should have all the required fields, be a signed input, etc."
+  (validator/and
+   [validator/input
+    validator/time-created
+    validator/full-diff]))
+
+(def edit-keys
+  (fn [ks]
+    (<> ks
+        (<> (keys input-keys)
+            [:diff-number :modified-at :author :signature]))))
+
+(def validator/diff-edit-input
+  "An input to edit an diff."
+  (validator/and
+   [validator/input
+    validator/time-modified
+    validator/diff-reference
+    (validator/contains-only (edit-keys (keys diff-keys)))]))
+
+(def strip-input
+  "After an entity input has been processed/verified, we no longer need some of
+  the fields."
+  (fn [e]
+  (delete-many [:chain-id :signature :nonce :diff-number] e)))
+
+(def create-diff
+  "Create an diff from a dict, checking that it is valid and storing it in `diffs`."
+  (fn [i]
+    (validator/diff-input i)
+    (def n (diff-counter :next))
+    (def i_
+      (strip-input
+       (<> i
+           {:number   n})))
+    (set-diffs (@ n) i_)
+    (mark-used-nonce (lookup :nonce i))))
+
+(def over-diffs
+  "A helper to modify stuff in the diff ref."
+  (fn [l f]
+    (over-ref diffs l f)))
+
+(def add-comment
+  "Add a comment to a diff."
+  (fn [c]
+    (validator/comment-input c)
+    (def c_ (strip-input c))
+    (over-diffs
+     (.. (@ (lookup :diff-number c)) (@ :comments))
+     (fn [cs] (add-right c_ cs)))
+    (mark-used-nonce (lookup :nonce c))))
+
+(def merge-new
```

Comment:
I'd say define a `flip` function in the prelude. Then you won't need this function and can instead just use `flip <>`

Reply:
As this is also used in rad-issues, I would address this outside of this PR.